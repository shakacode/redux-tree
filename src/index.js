export { createTree } from './createTree';
export { createBranch } from './createBranch';
export { createLeaf } from './createLeaf';
