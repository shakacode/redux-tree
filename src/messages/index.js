export const REDUCERS_EXAMPLES = `
    Provide action handlers to the "createLeaf" function in one of the following ways:

    Option #1:
      { ACTION_TYPE: (state, action) => state }

    Option #2:
      {
        ACTION_TYPE: {
          leaf: ['path', 'to', 'leaf'],
          reduce: (state, action) => state,
        },
      }

    Option #3:
      {
        ACTION_TYPE: [
          (state, action) => state,

          {
            leaf: ['path', 'to', 'leaf'],
            reduce: (state, action) => state,
          },
        ],
      }
`;


export const nonRecordState = state => {
  const stringifiedState = JSON.stringify(state);
  return `State must be Immutable Record. Received: ${stringifiedState}`;
};


export const undefinedState = (keyPath, action) => {
  const stringifiedKeyPath = JSON.stringify(keyPath);
  const stringifiedAction = JSON.stringify(action);

  return (`
    Reducer at keypath ${stringifiedKeyPath} returned undefined.
    To ignore an action, you must explicitly return the previous state.
    If you want this reducer to hold no value, you can return null instead of undefined.

    Keypath: ${stringifiedKeyPath}
    Action: ${stringifiedAction}
  `);
};


export const badKeyPath = (keyPath, action) => {
  const stringifiedKeyPath = JSON.stringify(keyPath);
  const stringifiedAction = JSON.stringify(action);

  return (`
    "action.${action.type}" was dispatched.
    Trying to update state at keypath ${stringifiedKeyPath} but resolved value is undefined.
    Most likely you misspelled one of the keys in the leaf keypath.
    Re-check your action handlers for this action.

    Keypath: ${stringifiedKeyPath}
    Action: ${stringifiedAction}
  `);
};


export const badActionHandler = (actionHandler, action) => {
  const stringifiedActionHandler = JSON.stringify(actionHandler);
  const stringifiedAction = JSON.stringify(action);

  return (`
    Invalid action handler is passed for "action.${action.type}".

    Received:
      Action: ${stringifiedAction}
      Action handler: ${stringifiedActionHandler}

    ${REDUCERS_EXAMPLES}
  `);
};
