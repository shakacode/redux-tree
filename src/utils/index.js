export { createShell } from './createShell';
export { isReduxAction } from './isReduxAction';
export { isPlainObject } from './isPlainObject';
