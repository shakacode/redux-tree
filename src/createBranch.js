import { Record } from 'immutable';

import { createShell, isReduxAction } from './utils';
import * as messages from './messages';

export const createBranch = children => {
  const childrenNames = Object.keys(children);

  let keyPath;

  return (state, action, parentKeyPath) => {
    if (!keyPath) {
      keyPath = [].concat(parentKeyPath);
    }

    return state.withMutations(nextState => {
      if (isReduxAction(action.type)) {
        const branchState = nextState.getIn(keyPath);

        if (!branchState) {
          nextState.setIn(keyPath, createShell(childrenNames));
        // Record.isRecord method is available only in immutable@4.x
        // For now skipping this check if older version is used
        } else if (Record.isRecord && !Record.isRecord(branchState)) {
          throw new Error(messages.nonRecordState(branchState));
        }
      }

      childrenNames.forEach(child => {
        const childKeyPath = keyPath.concat(child);
        // eslint-disable-next-line no-param-reassign
        nextState = children[child](nextState, action, childKeyPath);
      });
    });
  };
};
