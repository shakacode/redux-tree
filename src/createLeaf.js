import { isPlainObject } from './utils';
import * as messages from './messages';

export const createLeaf = (initialState, reducers) => {
  if (typeof initialState === 'undefined') {
    throw new Error(`
      "createLeaf" recieved undefined on initialisation.
      The initial state may not be undefined.
      If you don't want to set a value for this reducer,
      you can use null instead of undefined.
    `);
  }

  if (reducers && !isPlainObject(reducers)) {
    throw new Error(`
      Invalid action handlers are passed to "createLeaf" function.

      Received:
        arg #1: ${JSON.stringify(initialState)}
        arg #2: ${JSON.stringify(reducers)}

      ${messages.REDUCERS_EXAMPLES}
    `);
  }

  if (process.env.NODE_ENV !== 'production') {
    // TODO: Validate reducers shape.
  }

  return (state, action, keyPath, initialized) => {
    const leafKeyPath = Array.isArray(keyPath) ? keyPath : [keyPath];
    const leafState = state.getIn(leafKeyPath);

    return state.withMutations(nextState => {
      // If reducer is a plain function w/ the switch statement
      if (!reducers && typeof initialState === 'function') {
        const reducer = initialState;
        const nextLeafState = reducer(leafState, action);
        if (typeof nextLeafState === 'undefined') {
          throw new Error(messages.undefinedState(leafKeyPath, action));
        }
        return nextState.setIn(leafKeyPath, nextLeafState);
      }

      // On initialization
      if (!initialized) {
        // Do nothing if state is rehydrated
        if (typeof leafState !== 'undefined') return;
        // Otherwise set initial state
        return nextState.setIn(leafKeyPath, initialState);
      }

      // No reducers attached to the leaf,
      // it's fine as the state might be changed from another leafs
      if (!reducers) return;

      const { type, ...payload } = action;

      // No reaction required, exiting
      if (!reducers[type]) return;

      // If value of action type key is function -> reducing local leaf state
      if (typeof reducers[type] === 'function') {
        const nextLeafState = reducers[type](leafState, payload);
        if (typeof nextLeafState === 'undefined') {
          throw new Error(messages.undefinedState(leafKeyPath, action));
        }
        return nextState.setIn(leafKeyPath, nextLeafState);
      }

      // If it's an object w/ keypath to leaf and reducer -> reducing leaf state at keypath
      if (isPlainObject(reducers[type])) {
        const externalLeafKeyPath = reducers[type].leaf;
        const externalLeafState = nextState.getIn(externalLeafKeyPath);
        if (typeof externalLeafState === 'undefined') {
          // NOTE: Might worth to check if this is exactly leaf, not a branch, at provided keypath.
          //       But this data must be stored somewhere. Premature optimisation at this point.
          throw new Error(messages.badKeyPath(externalLeafKeyPath, action));
        }

        const nextExternalLeafState = reducers[type].reduce(externalLeafState, payload);
        if (typeof nextExternalLeafState === 'undefined') {
          throw new Error(messages.undefinedState(externalLeafKeyPath, action));
        }
        return nextState.setIn(externalLeafKeyPath, nextExternalLeafState);
      }

      // If it's an array -> iterating
      if (Array.isArray(reducers[type])) {
        return reducers[type].forEach(reducer => {
          // If array's item is a function -> reducing local leaf state
          if (typeof reducer === 'function') {
            const nextLeafState = reducer(leafState, payload);
            if (typeof nextLeafState === 'undefined') {
              throw new Error(messages.undefinedState(leafKeyPath, action));
            }
            return nextState.setIn(leafKeyPath, nextLeafState);
          }

          // Otherwise reducing state of the leaf at provided keypath
          const externalLeafState = nextState.getIn(reducer.leaf);
          if (typeof externalLeafState === 'undefined') {
            throw new Error(messages.badKeyPath(reducer.leaf, action));
          }

          const nextExternalLeafState = reducer.reduce(externalLeafState, payload);
          if (typeof nextExternalLeafState === 'undefined') {
            throw new Error(messages.undefinedState(reducer.leaf, action));
          }
          return nextState.setIn(reducer.leaf, nextExternalLeafState);
        });
      }

      throw new Error(messages.badActionHandler(reducers[type], action));
    });
  };
};
