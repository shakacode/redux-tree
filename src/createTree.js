import { Record } from 'immutable';

import { createShell } from './utils';
import * as messages from './messages';

export const createTree = children => {
  const childrenNames = Object.keys(children);

  let initialized = false;

  return (state = createShell(childrenNames), action) => {
    // Record.isRecord method is available only in immutable@4.x
    // For now skipping this check if older version is used
    if (state && Record.isRecord && !Record.isRecord(state)) {
      throw new Error(messages.nonRecordState(state));
    }

    return state.withMutations(nextState => {
      childrenNames.forEach(child => {
        // eslint-disable-next-line no-param-reassign
        nextState = children[child](nextState, action, child, initialized);
      });

      if (!initialized) initialized = true;
    });
  };
};
