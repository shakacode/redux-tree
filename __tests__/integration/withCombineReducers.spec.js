import { createStore, combineReducers } from 'redux';
import { Record, Set } from 'immutable';

import { createLeaf } from '../../src';
import { EntityLeaf, UILeaf, mockTree } from '../factories';

describe('redux-tree', () => {
  const rootReducer = combineReducers({
    counter: (state = 0, action) => {
      switch (action.type) {
        case 'INCREMENT_VANILLA_COUNTER':
          return state + 1;
        default:
          return state;
      }
    },
    tree: mockTree({
      entityLeaf: createLeaf(
        new EntityLeaf({ index: new Set() }),
        {
          ADD_ENTITY_TO_TREE:
            (state, { entityId }) =>
              state.update('index', index => index.add(entityId)),
        },
      ),
      uiLeaf: createLeaf(
        new UILeaf({ processingEntities: new Set() }),
        {
          START_PROCESSING_TREE_ENTITY:
            (state, { entityId }) =>
              state.update(
                'processingEntities',
                processingEntities => processingEntities.add(entityId),
              ),

          STOP_PROCESSING_TREE_ENTITY: [
            // 1. ui leaf
            (state, { entityId }) =>
              state.update(
                'processingEntities',
                processingEntities => processingEntities.delete(entityId),
              ),

            // 2. entity leaf
            {
              leaf: ['entities', 'entityLeaf'],
              reduce:
                (state, { entityId }) =>
                  state.update('index', index => index.delete(entityId)),
            },
          ],
        },
      ),
    }),
  });

  it('initializes empty state on the first dispatch w/o rehydrated state', () => {
    const store = createStore(rootReducer);
    const state = store.getState();

    expect(state.counter).toBe(0);
    expect(state.tree.entities).toBeInstanceOf(Record);
    expect(state.tree.entities.entityLeaf.index).toBeInstanceOf(Set);
    expect(state.tree.ui).toBeInstanceOf(Record);
    expect(state.tree.ui.uiLeaf.processingEntities).toBeInstanceOf(Set);
    expect(state).toMatchSnapshot('@@INIT');
  });


  it('initializes rehydrated state on the first dispatch', () => {
    // rehydrated state
    const State = Record({
      entities: undefined,
      ui: undefined,
    });

    const Entities = Record({
      entityLeaf: undefined,
    });

    const UI = Record({
      uiLeaf: undefined,
    });

    const rehydratedState = {
      counter: 42,
      tree: new State({
        entities: new Entities({
          entityLeaf: { index: new Set([1, 2, 3]) },
        }),
        ui: new UI({
          uiLeaf: { processingEntities: new Set() },
        }),
      }),
    };

    const store = createStore(rootReducer, rehydratedState);
    const state = store.getState();

    expect(state.counter).toBe(42);
    expect(state.tree.entities).toBeInstanceOf(Record);
    expect(state.tree.entities.entityLeaf.index).toBeInstanceOf(Set);
    expect(state.tree.ui).toBeInstanceOf(Record);
    expect(state.tree.ui.uiLeaf.processingEntities).toBeInstanceOf(Set);
    expect(state).toMatchSnapshot('@@INIT_WITH_STATE');
  });


  it('reduces the tree state', () => {
    const store = createStore(rootReducer);

    const initialState = store.getState();

    store.dispatch({
      type: 'ADD_ENTITY_TO_TREE',
      entityId: 1,
    });

    const stateWithEntity = store.getState();

    store.dispatch({
      type: 'START_PROCESSING_TREE_ENTITY',
      entityId: 1,
    });

    const stateWithProcessingEntity = store.getState();

    store.dispatch({
      type: 'STOP_PROCESSING_TREE_ENTITY',
      entityId: 1,
    });

    const stateWithoutEntity = store.getState();

    expect(initialState).toMatchSnapshot('@@INIT_TREE');
    expect(stateWithEntity).toMatchSnapshot('ADD_ENTITY_TO_TREE');
    expect(stateWithProcessingEntity).toMatchSnapshot('START_PROCESSING_TREE_ENTITY');
    expect(stateWithoutEntity).toMatchSnapshot('STOP_PROCESSING_TREE_ENTITY');
  });


  it('reduces vanilla counter state', () => {
    const store = createStore(rootReducer);

    store.dispatch({ type: 'INCREMENT_VANILLA_COUNTER' });

    const state = store.getState();

    expect(state).toMatchSnapshot('INCREMENT_VANILLA_COUNTER');
  });
});
