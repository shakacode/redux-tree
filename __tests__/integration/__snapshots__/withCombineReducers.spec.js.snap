// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@@INIT 1`] = `
Object {
  "counter": 0,
  "tree": Object {
    "entities": Object {
      "entityLeaf": Object {
        "index": Array [],
      },
    },
    "ui": Object {
      "uiLeaf": Object {
        "processingEntities": Array [],
      },
    },
  },
}
`;

exports[`@@INIT_TREE 1`] = `
Object {
  "counter": 0,
  "tree": Object {
    "entities": Object {
      "entityLeaf": Object {
        "index": Array [],
      },
    },
    "ui": Object {
      "uiLeaf": Object {
        "processingEntities": Array [],
      },
    },
  },
}
`;

exports[`@@INIT_WITH_STATE 1`] = `
Object {
  "counter": 42,
  "tree": Object {
    "entities": Object {
      "entityLeaf": Object {
        "index": Array [
          1,
          2,
          3,
        ],
      },
    },
    "ui": Object {
      "uiLeaf": Object {
        "processingEntities": Array [],
      },
    },
  },
}
`;

exports[`ADD_ENTITY_TO_TREE 1`] = `
Object {
  "counter": 0,
  "tree": Object {
    "entities": Object {
      "entityLeaf": Object {
        "index": Array [
          1,
        ],
      },
    },
    "ui": Object {
      "uiLeaf": Object {
        "processingEntities": Array [],
      },
    },
  },
}
`;

exports[`INCREMENT_VANILLA_COUNTER 1`] = `
Object {
  "counter": 1,
  "tree": Object {
    "entities": Object {
      "entityLeaf": Object {
        "index": Array [],
      },
    },
    "ui": Object {
      "uiLeaf": Object {
        "processingEntities": Array [],
      },
    },
  },
}
`;

exports[`START_PROCESSING_TREE_ENTITY 1`] = `
Object {
  "counter": 0,
  "tree": Object {
    "entities": Object {
      "entityLeaf": Object {
        "index": Array [
          1,
        ],
      },
    },
    "ui": Object {
      "uiLeaf": Object {
        "processingEntities": Array [
          1,
        ],
      },
    },
  },
}
`;

exports[`STOP_PROCESSING_TREE_ENTITY 1`] = `
Object {
  "counter": 0,
  "tree": Object {
    "entities": Object {
      "entityLeaf": Object {
        "index": Array [],
      },
    },
    "ui": Object {
      "uiLeaf": Object {
        "processingEntities": Array [],
      },
    },
  },
}
`;
